pkg/collector/pipeline.go

```diff
@@ -47,37 +50,79 @@ func (c nowClock) Now() time.Time {
 var defaultClock nowClock
 
 // Pipeline is a series of processors that should be applied to each report that
-// the collector receives.
+// the collector receives. It uses a fixed number of workers to process the reports
+// and a fixed sized queue that the workers read from. If the queue fills, reports
+// are dropped. Pipeline{} is not a usable instance, use NewPipeline for production
+// and NewTestPipeline* in tests.
 type Pipeline struct {
 	processors []ReportProcessor
 	clock      Clock
+	c          chan *ReportBatch
+	wg         *sync.WaitGroup
 }
 
-// NewPipeline creates a new Pipeline that uses a particular clock.  For
-// production pipelines, just instantiate the Pipeline type yourself
-// (&Pipeline{}).
-func NewPipeline(clock Clock) *Pipeline {
-	return &Pipeline{clock: clock}
+// NewPipeline creates a new Pipeline with a specified buffer size.
+func NewPipeline(bufferSize int64, numWorkers int) *Pipeline {
+	return setupPipeline(context.Background(), nil, bufferSize, numWorkers)
+}
+
+const defaultBufferSize = 1000
+const defaultNumWorkers = 10
+
+// NewTestPipeline creates a new Pipeline with a specified clock.
+// This should only be used for testing.
+func NewTestPipeline(clock Clock) *Pipeline {
+	return setupPipeline(context.Background(), clock, defaultBufferSize, defaultNumWorkers)
```

Comment:
nit: I would make this call NewTestPipeline with defaultBufferSize as an argument

Reply:
Done